include/zephyr/drivers/charger.h

```diff
@@ -0,0 +1,182 @@
+	/** External supply is present and of fixed output */
+	CHARGER_ONLINE_FIXED,
+	/** External supply is present and of programmable output*/
+	CHARGER_ONLINE_PROGRAMMABLE,
+};
+
+/**
+ * @brief Charging states
+ */
+enum charger_status {
+	/** Charging device state is unknown */
+	CHARGER_STATUS_UNKNOWN = 0,
+	/** Charging device is charging a battery */
+	CHARGER_STATUS_CHARGING,
+	/** Charging device is not able to charge a battery */
+	CHARGER_STATUS_DISCHARGING,
+	/** Charging device is not charging a battery */
+	CHARGER_STATUS_NOT_CHARGING,
+	/** The battery is full and the charging device will not attempt charging */
+	CHARGER_STATUS_FULL,
+};
+
+/**
+ * @typedef charger_property_t
+ * @brief A charger property's identifier
+ *
+ * See charger_props for a list of identifiers
+ */
+typedef uint16_t charger_property_t;
+
+/**
+ * @brief Representation of a charger property value
+ */
+union charger_value {
+	enum charger_online online;
+	bool present;
+	enum charger_status status;
+	int custom_val;
+};
+
+/**
+ * @typedef charger_value_t
+ * @brief A type representing charger_value
+ */
+typedef union charger_value charger_value_t;
+
+/**
+ * @typedef charger_get_property_t
+ * @brief Callback API for getting a charger property.
+ *
+ * See charger_get_property() for argument description
+ */
+typedef int (*charger_get_property_t)(const struct device *dev, charger_property_t prop,
+				      charger_value_t *val);
+
+/**
+ * @brief Charging device API
+ *
+ * Caching is entirely on the onus of the client
+ */
+__subsystem struct charger_driver_api {
+	charger_get_property_t get_property;
+};
+
+/**
+ * @brief Fetch a battery charger property
+ *
+ * @param dev Pointer to the battery charger device
+ * @param prop pointer to charger_get_property struct where the property struct
+ * field is set by the caller to determine what property is read from the
+ * charger device into the charger_get_property struct's value field.
+ *
+ * @retval 0 if successful
+ * @retval < 0 if getting property failed
+ */
+__syscall int charger_get_prop(const struct device *dev, charger_property_t prop,
+			       charger_value_t *val);
+
+static inline int z_impl_charger_get_prop(const struct device *dev, charger_property_t prop,
+					  charger_value_t *val)
+{
+	const struct charger_driver_api *api = (const struct charger_driver_api *)dev->api;
+
+	return api->get_property(dev, prop, val);
+}
+
+/**
+ * @brief Fetch multiple battery charger properties
+ *
+ * @param dev Pointer to the battery charger device
+ * @param props pointer to array of charger_get_property struct where the property struct
+ * field is set by the caller to determine what property is read from the
+ * charger device into the charger_get_property struct's value field. The array is not
+ * permuted and elements can be directly accessed as loaded.
+ * @param props_len number of properties in props array
+ *
+ * @retval 0 if successful
+ * @retval < 0 if getting a property failed
```

Comment:
nit: Would prefer if this was non-zero if getting a property failed. In case we do something in the future like using a bit-mask or positive numbers to indicate more information about the number or the wherabouts of the failures. I don't think this is worth holding up the MVP experimental API PR though.